### ISSUE TITLE
Simplifying the sRGB and colorspace situation

### ISSUE THREAD

**Issue by jasoneveleth** [glium/glium#2069]
This issue is to follow up my previous issue #2059.

As it turns out, a lot of the implicit conversions are coming from OpenGL itself. See my [blog post](https://www.jasoneveleth.com/blog/2023/07/31/understanding-GLFRAMEBUFFERSRGB-in-glium). I summarized all the issues people have had with glium and colors. I made a flowchart (which is handwritten for now) which captures all the fixes. I also outlined two possible approaches to make using glium easier. I'll cover them here for convenience.

We can either hide the fact that OpenGL uses `GFS`, or we should be very transparent about what `GFS` means and how to interact with it.

## Transparent Approach

Here, `GFS` should be enabled or disabled when the context is initialized and never changed (there is already support for this in `ContextBuilder`). Using `clear_color()` to change the `GFS` value seems like the wrong way to go since it introduces unnecessary complexity. I can't think of a use case where being able to change `GFS` at runtime is useful if we're trying to be transparent (feel free to point one out in the comments).

In this approach with `GFS` enabled, we still require users to use `SrgbTexture`s and `Texture`s, and to differentiate what type of values are stored there. But, there is no need to specify whether a program outputs sRGB, since we are going to use OpenGL's underlying implicit conversions.

In this approach, if you choose to disable `GFS`: 
- you should do everything using linear RGB and no gamma correction will be applied (except from `SrgbTexture`, which will of course be converted to linear when sampled from)
- we assume the user understands what they are doing

### How to implement this approach

First, remove `clear_color_srgb()` and don't fiddle with `GFS` in `clear_color()`. Second, remove the code which enables `GFS` in `use_program()` at `5d50e70:src/program/program.rs:449`.

### Advantages of this approach

- familiarity: it respects the OpenGL model (using `GFS` as a global parameter) and obeys the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) 
- easy-of-use for beginners assuming they already understand OpenGL

### Disadvantages

- blending with an sRGB image won't work nicely with `GFS` disabled, so all data must be in linear color space, which is mildly inconvenient

## Hidden approach

The user should never have to worry about `GFS`. This is most similar to how things are done now. They do still need to worry about whether their textures and programs contain/output sRGB values.

### How to implement this approach

We shouldn't let users choose the `GFS` setting when initializing a context since they shouldn't worry about it. And `GFS` should always be enabled. We will still need users to tell us whether they're texture has linear or sRGB values, and whether their program outputs linear or sRGB. But, that is the only time they'll need to worry about colorspace, we'll handle the enabling/disabling of `GFS` behind the scenes (similar to how we do it in `use_program()` at `5d50e70:src/program/program.rs:449`).

We also should remove `clear_color_srgb()` since users shouldn't worry about `GFS`.

### Advantage of this approach

- simplicity: there's only 1 mode to worry about
- backward compatibility with old `glium` programs since this is doesn't change the fundamental model

**Comment by timlathy** [comment]
Given that this crate is in maintenance mode, I'm not sure that radical changes such as removing runtime control over `GL_FRAMEBUFFER_SRGB` are good. I'd say I prefer the second approach, but given that `clear_color` and `clear_color_srgb` produce different results, why not keep both? (That being said, perhaps their behavior [should be reversed](https://github.com/glium/glium/issues/2059#issuecomment-1666613558) to match how `outputs_srgb` works in Program.)

Also, what do you think about defaulting `outputs_srgb` to true in Program, [as suggested](https://github.com/glium/glium/issues/2059#issuecomment-1642216761) by @Melchizedek6809? I think the current behavior of enabling `GL_FRAMEBUFFER_SRGB` by default is counterintuitive, and fixing it in your code is almost like a rite of passage for getting started with Glium — like [you and other people](https://github.com/glium/glium/issues/2059#issuecomment-1642595208), I also had to spend some time going through the docs and GitHub issues before I realized I need to create my Programs with `outputs_srgb: true`.

To sum up, I'd suggest starting with two simple changes:
1. Swap `clear_color` and `clear_color_srgb`
2. By default, set `outputs_srgb` to true in Program

Though it doesn't really fix the conceptual issues, I think it should at least result in less unpleasant surprises for beginners.

**Comment by est31** [comment]
@Melchizedek6809 btw, changing the default of `outputs_srgb` to true would be a good idea, I agree.

The swapping suggestion basically amounts to a revert of 047cb0104a67b2a038c2c89a5994918f780c927c , which I would also be okay with.

I remember running into srgb issues myself when I used glium. Note though that if we do the `outputs_srgb` change, we will probably disrupt users. I think that's okay though, glium is not stable and the issue is fairly noticeable (although a compile error would be nicer).
